### ISSUE TITLE
no-implicit-coercion: Separate options for !!x and ~x.indexOf("y") [$15]

### ISSUE THREAD

**Issue by cvrebert** [eslint/eslint#3308]
Currently, the `boolean` option of `no-implicit-coercion` controls both of these. I think it would be nice to be able to control/(dis)allow these separately.
Personally, I find `!!` tolerable but the `~` hack too obscure.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by nzakas** [comment]
Do you have a proposal for how this could work?

**Comment by cvrebert** [comment]
I suppose, for backward compatibility, allow `boolean` to accept *either* a single boolean or an object as its value. The new object would be something like (using strawman names) `{"not-not": true, "bitwise-not": false}`, thus allowing control over `!!x` and `~x.indexOf("y")` (respectively).
Setting `boolean` to a boolean value would set both of these sub-settings to the same value, thus retaining backward compatibility.

**Comment by awerlang** [comment]
I agree to this proposed solution.

**Comment by nzakas** [comment]
Interesting idea, though I think perhaps using the actual character y `!!` and `~` might be clearer?

**Comment by cvrebert** [comment]
Sure, sounds fine. My names were just examples; I'n not attached to them at all.

**Comment by nzakas** [comment]
So what we're looking at is a configuration that looks like:

```
no-implicit-coercion: [2, {
    boolean: {
        "!!": false,
        "~": false
    }
}
```

Looking at it again, I think this is confusing. So maybe this instead:

```
no-implicit-coercion: [2, {
    exceptions: ["!!", "~"],
    boolean: true
}
```

Thoughts?

**Comment by cvrebert** [comment]
> Looking at it again, I think this is confusing.

How so?

**Comment by nzakas** [comment]
Because:

1. You'd need to know that these are boolean coercions
2. You'd then expect to be able to do the same in the other sections
3. How would you set the default for `boolean`?

**Comment by ilyavolodin** [comment]
I think we should go with `exceptions` to stay consistent with other rules.

**Comment by nzakas** [comment]
Or `allow` to really go in the direction we're heading with other rules?

**Comment by ilyavolodin** [comment]
Yes, `allow` is probably better in this case. As in allow this pattern.

**Comment by rpatil26** [comment]
Let me know if anyone else has already picked it. If not, I am going to work on it tonight.

**Comment by nzakas** [comment]
Yes, we should allow overrides for any of the operators we currently flag.

**Comment by rpatil26** [comment]
How do we specify `+` differently, one for `"" + stringVariable` case and other for ` +booleanVariable` case?

**Comment by rpatil26** [comment]
I am done with my changes except that empty string contact `+` operator overridability.

Example configuration that allows `!!`:
```js
no-implicit-coercion: [2, {
    boolean: true,
    string: true,
    number: true,
    allow: ["!!", /*"~", "*", "+"*/]
}
```

**Comment by rpatil26** [comment]
@nzakas Please let me know so that I can move on and create PR. I couldn't come up with any reliable way to specify unary and binary + separately.

**Comment by nzakas** [comment]
We probably need to allow all uses of `+` and just not worry about differentiating. Just make a note of it in the docs.
